apis/r/src/rinterface.cpp

```diff
@@ -624,6 +624,66 @@ Rcpp::List c_attributes(const std::string& uri, Rcpp::XPtr<somactx_wrap_t> ctxxp
     return result;
 }
 
+// adapted from tiledb-r
+// https://github.com/TileDB-Inc/TileDB-R/blob/525bdfc0f34aadb74a312a5d8428bd07819a8f83/src/libtiledb.cpp#L3027-L3042
+// [[Rcpp::export]]
+Rcpp::LogicalVector c_attribute_enumerated(const std::string& uri, Rcpp::XPtr<somactx_wrap_t> ctxxp) {
+    auto sr = tdbs::SOMAArray::open(OpenMode::read, uri, ctxxp->ctxptr);
+    std::shared_ptr<tiledb::ArraySchema> sch = sr->tiledb_schema();
+    sr->close();
+
+    int nattrs = sch->attribute_num();
+    Rcpp::LogicalVector has_enum = Rcpp::LogicalVector(nattrs);
+    Rcpp::CharacterVector names = Rcpp::CharacterVector(nattrs);
+    for (int i = 0; i < nattrs; i++) {
+        //
+        auto attr = make_xptr<tiledb::Attribute>(new tiledb::Attribute(sch->attribute(i)));
+        auto enmr = tiledb::AttributeExperimental::get_enumeration_name(
+            // credit to this monstrosity goes to Beka Davis
+            const_cast<const tiledb::Context&>(*(ctxxp.get()->ctxptr.get()->tiledb_ctx())),
+            *attr.get()
+        );
+        has_enum(i) = enmr != std::nullopt;
+        names(i) = attr->name();
+    }
+
+    has_enum.attr("names") = names;
+    return has_enum;
+}
+
+// [[Rcpp::export]]
+Rcpp::CharacterVector c_attribute_enumeration_levels(
+        const std::string& uri,
+        Rcpp::XPtr<somactx_wrap_t> ctxxp,
+        const std::string& name
+) {
+    auto sr = tdbs::SOMAArray::open(OpenMode::read, uri, ctxxp->ctxptr);
+    auto enum_values = sr->get_enumeration_values_for_column(name);
+    sr->close();
+
+    if (enum_values.first->length > std::numeric_limits<int32_t>::max()) {
+        Rcpp::stop("too many enumeration levels for R");
+    }
+
+    nanoarrow::UniqueArrayView enum_view;
+    ArrowArrayViewInitFromType(enum_view.get(), NANOARROW_TYPE_STRING);
```

Comment:
```suggestion
    ArrowArrayViewInitFromType(enum_view.get(), NANOARROW_TYPE_LARGESTRING);
```

https://github.com/single-cell-data/TileDB-SOMA/blob/1.16.1/libtiledbsoma/src/external/include/nanoarrow/nanoarrow.h#L3689-L3701

since core uses (always) 64-bit offsets